### ISSUE TITLE
[MDEP-753] Non-test dependency reported as Non-test scoped test only dependency

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-dependency-plugin#1254]
**[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo)** opened **[MDEP-753](https://issues.apache.org/jira/browse/MDEP-753?redirect=false)** and commented

Saw this when updating the google-http-java-client from 3.1.2 to 3.2.0 of the plugin. I'm not immediately sure whether this is a regression:

[INFO] --- maven-dependency-plugin:3.2.0:analyze (default-cli) @ google-http-client ---
Warning:  Non-test scoped test only dependencies found:
Warning:     com.google.guava:guava:jar:30.1.1-android:compile
Warning:     io.opencensus:opencensus-api:jar:0.28.0:compile

Changing Guava to scope test breaks the build, which is expected based on the code. The warning seems incorrect.

https://github.com/googleapis/google-http-java-client/pull/1396
https://github.com/googleapis/google-http-java-client/pull/1396/checks?check_run_id=2809438131

---

**Affects:** 3.2.0

**Attachments:**
- [chas.zip](https://issues.apache.org/jira/secure/attachment/13039958/chas.zip) (_1.69 kB_)
- [tj.zip](https://issues.apache.org/jira/secure/attachment/13039955/tj.zip) (_1.40 kB_)

**Issue Links:**
- [MSHARED-1021](https://issues.apache.org/jira/browse/MSHARED-1021) Handle different classes from same artifact used by model and test code
 (_**"is blocked by"**_)
- [MDEP-784](https://issues.apache.org/jira/browse/MDEP-784) Upgrade maven-dependency-analyzer to 1.12.0
 (_**"is blocked by"**_)
- [MDEP-791](https://issues.apache.org/jira/browse/MDEP-791) Non-test scoped and transitive dependencies in compile scope
 (_**"split to"**_)
- [MDEP-790](https://issues.apache.org/jira/browse/MDEP-790) New Goals: analyze-main and analyze-test
 (_**"split to"**_)
- [MDEP-757](https://issues.apache.org/jira/browse/MDEP-757) bug with "non-test scoped test only dependencies found"

- [MDEP-759](https://issues.apache.org/jira/browse/MDEP-759) 'Dependency not found' with 3.2.0 and Java-17 while analyzing

- [MDEP-780](https://issues.apache.org/jira/browse/MDEP-780) “Non-test scoped test only dependencies found” false-positive in 3.2.0/1.11.2

- [MDEP-754](https://issues.apache.org/jira/browse/MDEP-754) Analyze incorrectly warns that junit is test only when it is a compile time dependency

- [MDEP-787](https://issues.apache.org/jira/browse/MDEP-787) Allow ignoring non-test-scoped dependencies

**Remote Links:**
- [GitHub Pull Request #34
](https://github.com/apache/maven-dependency-analyzer/pull/34)
- [GitHub Pull Request #35
](https://github.com/apache/maven-dependency-analyzer/pull/35)
- [GitHub Pull Request #37
](https://github.com/apache/maven-dependency-analyzer/pull/37)

40 votes, 45 watchers

**Comment by jira-importer** [comment]
**[Charles Honton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=charles_honton@intuit.com)** commented

[^chas.zip]

Added analyze-only execution to the pom.

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

[chonton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chonton) - please not include more cases here - this issue is closed

**Comment by jira-importer** [comment]
**[Charles Honton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=charles_honton@intuit.com)** commented

In my opinion, the latest zip demonstrates that this issue still exists and that this issue should be re-opened.

**Comment by jira-importer** [comment]
**[Charles Honton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=charles_honton@intuit.com)** commented

This changes the semantics of 'compile' scope.  The definition of a parent class or interface is required to compile a class.  The parent and list of interfaces are available in the byte code.

Having a new option will require additional, per module, changes to poms.  A new version should just drop in without requiring a change in the pom.

**Comment by jira-importer** [comment]
**[Thorsten Glaser](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mirabilos)** commented

This is a little tricky.

Some of the other checks also check if the code **directly** uses the Maven module, and if not (if only subclasses that are in other modules are used), there normally needs to be no direct dependency from your project to the module with the parent classes that are not used.

That being said, you say “with test scope, compile fails”. Well, it does not, for me:

```
$ mvn clean test
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------------------------< mdep753:tstp >----------------------------
[INFO] Building tstp 0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ tstp ---
[INFO] Deleting /home/tglase/tjc/target
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ tstp ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/tglase/tjc/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ tstp ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /home/tglase/tjc/target/classes
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ tstp ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/tglase/tjc/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ tstp ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /home/tglase/tjc/target/test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ tstp ---
[INFO] Surefire report directory: /home/tglase/tjc/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running tstp.TestMDEP753
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.512 sec

Results :

Tests run: 1, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.025 s
[INFO] Finished at: 2022-02-13T19:49:03+01:00
[INFO] ------------------------------------------------------------------------
```

This is with `chas.zip` with only the following change:

```
--- tj/pom.xml
+++ tjc/pom.xml
@@ -45,8 +45,8 @@
                        <groupId>com.fasterxml.jackson.core</groupId>
                        <artifactId>jackson-core</artifactId>
                        <version>${jackson.version}</version>
-<!--   <scope>test</scope>
-                       with test scope, compile fails
+       <scope>test</scope>
+<!--                   with test scope, compile fails
                        without test scope, dependency check fails
 [INFO] - - - maven-dependency-plugin:3.3.0-SNAPSHOT:analyze-only (check-dependencies) @ tstp - - -
 [WARNING] Non-test scoped test only dependencies found:
```

What do you do differently that it fails “mvn clean test” for you? In case this is relevant, I’m using:

```
$ mvn --version
Apache Maven 3.6.3
Maven home: /usr/share/maven
Java version: 11.0.14, vendor: Debian, runtime: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_GB, platform encoding: UTF-8
OS name: "linux", version: "5.10.0-8-amd64", arch: "amd64", family: "unix"
```

If it failed, I’d agree that this was still unfixed / a bug, but `mvn clean test` (and `{}package{`}) passing is somewhat of the litmus test here… and they do, for me.

**Comment by jira-importer** [comment]
**[Sergey Vyacheslavovich Brunov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sbrunov)** commented

Dear [Thorsten Glaser](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mirabilos),

> [Sergey Vyacheslavovich Brunov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sbrunov) how did you create the box for the preformatted output? Mine ↑ here looks awful ☹

It seems that you have already found how to do it.
Answering for the sake of completeness.
To make a block of source code preformatted, surround it with {code}.

For additional details, please, refer to the page: [Text Formatting Notation Help - ASF JIRA](https://issues.apache.org/jira/secure/WikiRendererHelpAction.jspa?section=all).

Best regards,
Sergey Vyacheslavovich Brunov.

**Comment by jira-importer** [comment]
**[Henning Schmiedehausen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henning)** commented

The current 3.3.0-SNAPSHOT has some serious regressions/problem. It is not fit to be released. Here is a very obvious one (and easy to reproduce):

```
% git clone git@github.com:jdbi/jdbi.git
% cd jdbi
% git checkout -b test v3.27.1
% mvn -DskipTests clean install -pl :jdbi3-spring5
```

This builds fine. It uses the 3.1.2 dependency plugin; the last one that actually works.

with the current 3.3.0-SNAPSHOT (and the 1.12.0 dependency analyzer library):

```
% mvn -DskipTests clean install -pl :jdbi3-spring5 -Ddep.plugin.dependency.version=3.3.0-SNAPSHOT -Ddep.dependency-analyzer.version=1.12.0
[...]
[WARNING] Non-test scoped test only dependencies found:
[WARNING]    org.springframework:spring-core:jar:5.3.12:compile
```

Clearly, there are spring-core dependencies all over this project. Putting this dependency into test scope fails the build with compile errors.

I feel all of the post-3.1.2 work (and post 1.11.1 dependency analyzer) needs to be revisited. There are a number of very serious regressions that have not been addressed for a very long time.

**Comment by jira-importer** [comment]
**[Thorsten Glaser](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mirabilos)** commented

Thanks [hgschmie](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hgschmie), this is ridiculous but I **think** I understand the bug now.

Trying to build as-is will fil with one “non-test scoped test only” (and this is probably missing a few hyphens) warning, because…

```
$ git grep -F org.springframework.core
spring5/src/test/java/org/jdbi/v3/spring5/TestVersion.java:import org.springframework.core.SpringVersion;
```

… only the test ever imports **directly** from `{}spring-core{`}. However, you **cannot** switch the dependency to `test` scope because it’s needed as parent class:

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /tmp/jdbi/spring5/src/main/java/org/jdbi/v3/spring5/JdbiUtil.java:[65,20] cannot access org.springframework.core.Ordered
  class file for org.springframework.core.Ordered not found
[INFO] 1 error
[INFO] ------------------------------------------------------------- 
```

Line 65 is:

```
$ sed -n 65p /tmp/jdbi/spring5/src/main/java/org/jdbi/v3/spring5/JdbiUtil.java
    private static class Adapter extends TransactionSynchronizationAdapter { 
```

So it’s probably a parent class of `TransactionSynchronizationAdapter` that cannot be found.

But you cannot just remove the dependency (and rely on the implicit dependencies from the other Spring Maven modules) either:

```
[INFO] --- maven-dependency-plugin:3.3.0-SNAPSHOT:analyze-only (basepom.default) @ jdbi3-spring5 ---
[WARNING] Used undeclared dependencies found:
[WARNING]    org.springframework:spring-core:jar:5.3.12:compile
[WARNING] Non-test scoped test only dependencies found:
[WARNING]    org.springframework:spring-core:jar:5.3.12:compile 
```

The “used undeclared” is right here because the test **directly** uses it.

---

Yet, I think the behaviour of the 3.3.0-SNAPSHOT version of the maven-dependency-plugin is correct here, you’ve just reached a corner case that cannot be modelled properly within the bounds of the Maven model. In this case, I believe an `ignoredNonTestScopedDependencies` entry to override is the proper solution.

The thing here is that maven-dependency-plugin has two conflicting goals. One is that only modules that are directly used are listed as dependencies; recursive dependencies should take care of the rest. The other is that all classes that are directly used **are** indeed listed as dependencies. But while you can add a module in two different scopes as dependency (e.g. `test` and `{}provided{`}) you cannot add it to `test` scope explicitly but rely on implicit inclusion into `compile` scope.

It might be possible to extend the maven-dependency-plugin to somehow handle this case in a more user-friendly way, but this is literally a situation that _cannot_ be properly modelled in the project itself, so I’d think papering over this, while reducing angry users’ outcry, will do a disservice in the long run.

Maybe the new “non-test-scoped test-only dependencies” check should be a nōnfatal warning for a release or two? So projects continue to build, and people can add `ignoredNonTestScopedDependencies` entries at their leisure.

Maybe the plugin or the analyser can recognise this situation (specifically 「a Maven module is used (in `compile` scope) indirectly by other modules depending on it by means of inheritance, and at least one of these other modules (e.g. `{}spring-beans{`}) is indeed directly used by `main` classes, but the module is **directly** used by `test` classes」, this is what I believe to be the problem spec) and recommend adding `ignoredNonTestScopedDependencies` entries (it could provide them as copy/paste templates, even!) documenting this as a modelling problem.

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

[hgschmie](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hgschmie) it is the same problem, like in example from [chonton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chonton) 

 

In module `spring5` it is only one usage of `spring-core` in `org.jdbi.v3.spring5.TestVersion` another production code don't use `spring-core` ... but is is obviously that it is needed in compile scope.

**Comment by jira-importer** [comment]
**[Henning Schmiedehausen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henning)** commented

from what I can see is that "jdbi3-spring5 -> compile scope -> some-other-spring-dep -> compile scope -> spring-core" is needed to **build** the artifact.  But "jdbi3-spring5 -> test-scope -> spring-core" is actually the only place where the module explicitly references the module.

The maven core dependency resolver requires the the artifact to be in compile scope. So it can **not** be declared as "test' scope, because otherwise the core dependency resolver omits it from the compilation classpath and compilation fails. That is how the **core** resolver works.

What you basically tell me is "oops, dependency plugin thinks different, you need to work around the core resolver behavior. Tough luck".

The role of the dependency plugin is to report and check **what the core does**. Not what any other developer **thinks** it is right. The 3.1.2 version of this plugin does that. It clearly states that **if you put the spring-core dependency in the pom, it MUST be in compile scope because that is what the core resolver expects**.

What you are saying is that "well, you put it in compile scope, but we think it is wrong, because it is only used in test scope. However, you can not put it in test scope, because then the core resolver does not behave how we think it should. So, just exclude them from checking".

This is literally the opposite of what a user would expect from this plugin. For years, the dependency plugin has done exactly that: reflected what the core resolver does and reported accordingly. Changing this behavior now and telling users "tough luck, you need to change your configurations" is the opposite of what a minor (!) update (3.1.x -> 3.3.x) should do. You **may** get away with it for a 4.x.x. release but I would suggest starting a new plugin with that behavior.

Please, fix the bugs that plague 3.1.x, not change its behavior. Looking at the comments on this thread, you are on the wrong path. None of your users agrees that what you are doing is the right thing to do.

**Comment by jira-importer** [comment]
**[Henning Schmiedehausen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henning)** commented

> Maybe the new “non-test-scoped test-only dependencies” check should be a nōnfatal warning for a release or two? So projects continue to build, and people can add ignoredNonTestScopedDependencies entries at their leisure.

I am sorry, but you are wrong. Flagging transitive compile dependencies as "non-test scoped test-only dependency" is incorrect, because these dependencies are **not** test-only. They are required by the compiler to compile the code. They may not be explicit but if Maven **had** a way to turn off transitive resolution, they would need to be listed as part of the compile class path. By listing them in test scope, maven actually does two things in the core resolver:

- adding them to the test class path
- if they were a transitive dependency on the compile class path, then it **removes** them from the compile class path

The second part is what the 3.3.0-SNAPSHOT version of the dependency plugin currently gets wrong. And it is the behavior of the **core resolver**. We may disagree with that behavior (maybe the core resolver **should** keep them as transitive compile dependency even if they are listed as "test") but for now, **this is what it does**. And the dependency plugin **must** match the core behavior, not some other policy. We have been through that pain in the maven 2 cycle, I dearly do not want to go back there.

The current core resolver does not allow me to list these deps in "test" because it would fail the compilation. But the dependency plugin **requires** me to list the dependency because it is **used** in the code base (in the test classes).

- I need to declare the dependency
- I can not declare it in "test" scope

The correct answer for all versions of the plugin was/is "list as compile dependency" because that is the least harm. We are now requiring a dependency that is required anyway as part of the transitive dependency graph. Flagging this as "non-test scoped test-only dependency" is incorrect because this is a "direct test, transitive compile dependency".

The dependency plugin exists to allow users to validate that their pom configurations correctly reflect the dependencies used to build their code. The 3.1.2 plugin does that. The 3.2.0 and the 3.3.0-SNAPSHOT versions do not.

**Comment by jira-importer** [comment]
**[Thorsten Glaser](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mirabilos)** commented

> Flagging transitive compile dependencies as "non-test scoped test-only dependency" is incorrect, because these dependencies are **not** test-only.

This is not correct.

> They are required by the compiler to compile the code.

This is correct. But that doesn’t mean they are not test-only: only the tests **directly** use it.

As far as I can tell, maven-dependency-plugin here attempts to do something that is not currently possible with Maven. OpenBSD’s change to binutils (hide transitive dependencies in the `ld(1)` linker if they are not also explicitly depended on) is something I have long wished for, but it doesn’t work with the contemporary Java™ model of just throwing all dependencies into “the” classpath.

Rather, the compiler would need to see **only** the explicitly depended-on modules (so that using a symbol from e.g. spring-core in `src/main/` when not explicitly depended on would be a compiler failure!), and only for looking _into_ +these+ it would even consider _their_ transitive _direct_ dependencies… and for those, theirs, etc.

I’m all for minimising the direct dependencies, of course.

In my last comment I suggested the plugin recognise the very situation in your example and suggest proper ignore XML. Another option would be to just skip the warnings in that case (direct dependency only in `src/test/` but indirect dependency in `compile` scope from a module that’s directly depended on (this part is important)). It’d be a realistic compromise.

One of the developers said the test-only analysis only looks at bytecode. But I believe this can be done in a kind of postprocessing step. When we have a list of test-only direct dependencies, and we can obtain a list of both direct and transitive dependencies in nōn-test scope, and if a test-only dependency shows up in the latter list, the warning is skipped. (Maybe it’s not that easy. Also, to get the scopes right, there’s more than two, might be tricky. But it’d be worthwhile.)

**Comment by jira-importer** [comment]
**[Henning Schmiedehausen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henning)** commented

Thorsten,

I really appreciate that you look into this problem and that you are trying to solve it. However, you keep looking at it the wrong way:

- This is not OpenBSD. I know that you use that as an example but it does not change anything. They can do whatever they want to their users, it may be an interesting data point but should not serve as an example ("Starbucks no longer gives free cups to customers, so Peet's can not do that either" is not an argument that holds water for me)
- the dependency plugin has an understood behavior for many years (15 to be exact). Changing this now as part of a "3.1 -> 3.x" transition is wrong and the wrong approach to users
- you are arguing that the "current behavior is wrong and needs to be changed". What you experience right now is Hyrum's law (hi [hwright](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hwright)). The current behavior is what users expect and changing it leads to user pushback.

those are the problems. What I suggest to make this a successful transition:

- put these new checks under a switch ("enableStrictDependencyChecks"), turn it off by default. Release. That will make all the difference to the people stuck on 3.1.2 right now
- turn the switch on by default in a later release. People can still go back by saying "I don't want that".
- work with the core maven team to change the core resolver behavior. Make the dependency plugin match that. This is a long game (but then again we do have three different dependency resolvers already, why not have a fourth?)

But what you are doing right now is **user unfriendly**. And that is IMHO not the right thing to do.

**Comment by jira-importer** [comment]
**[Thorsten Glaser](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mirabilos)** commented

You seem to be confusing me with the developers and ignoring the proposed solution.

Separating the new checks into a new configuration option also sounds like a sensible thing to do but is orthogonal to that. (I even agree with that doing it as new, not yet enabled-by-default, option is the way to go for minor releases; unfortunately, Maven plugins seem to occasionally break things even in minor releases so I was not daring voicing it, plus it doesn’t fix the actual problems with this.)

**Comment by jira-importer** [comment]
**[Charles Honton](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=charles_honton@intuit.com)** commented

As a user of the dependency plugin, I expect the analyze and analysis-only goals to implement the following algorithm.

- Use byte code inspection on each class in the main and test artifacts to determine the set of classes required to compile the main and test sources.  These classes are the used-classes set.
- Determine in which dependencies each used class is present.  Add these dependencies to the used-dependency set.
- Add any dependency declared in the usedDependencies parameter to the used-dependency set.
- If ignoreNonCompile parameter is true, the declared-dependency set is the compile-scope dependencies from the maven resolver.  Else, if ignoreUnusedRuntime parameter is true, the declared-dependency set is the compile-scope, provided-scope, and system-scope dependencies from the maven resolver. Otherwise, the declared-dependency set is the compile-scope, provided-scope, system-scope, and test-scope dependencies from the maven resolver.
- Add each dependency in the used-dependency set that is not in the declared-dependency set to the used-but-undeclared set.
- Remove any dependency from the used-but-undeclared set that matches patterns declared in the  ignoreUsedUndeclaredDependencies parameter or the ignoreDependencies parameter.
- Add each dependency in the declared-dependency that is not in the used-dependency set to the declared-but-unused set.
- Remove any dependency from the declared-but-unused set that matches patterns declared in the ignoreUnusedDeclaredDependencies parameter or ignoreDependencies parameter.
- Print remaining members of used-but-undeclared set to the console or xml file.
- Print remaining members of declared-but-unused set to the console.
- If failOnWarning parameter is true, fail the build

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

I've splitted reported problem to new issue please look MDEP-791
